flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java

```diff
@@ -0,0 +1,156 @@
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static org.apache.flink.runtime.checkpoint.CheckpointFailureReason.MINIMUM_TIME_BETWEEN_CHECKPOINTS;
+import static org.apache.flink.runtime.checkpoint.CheckpointFailureReason.TOO_MANY_CONCURRENT_CHECKPOINTS;
+
+@SuppressWarnings("ConstantConditions")
+class CheckpointRequestDecider {
+	private static final int MAX_QUEUED_REQUESTS = 1000;
+
+	private final int maxConcurrentCheckpointAttempts;
+	private final Consumer<Long> rescheduleTrigger;
+	private final Clock clock;
+	private final long minPauseBetweenCheckpoints;
+	private final Supplier<Integer> pendingCheckpointsSizeSupplier;
+	private final Function<Supplier<Optional<CheckpointTriggerRequest>>, Optional<CheckpointTriggerRequest>> sync;
+	private final PriorityQueue<CheckpointTriggerRequest> triggerRequestQueue = new PriorityQueue<>((r1, r2) -> {
+		if (r1.props.isSavepoint() != r2.props.isSavepoint()) {
+			return r1.props.isSavepoint() ? -1 : 1;
+		} else if (r1.props.forceCheckpoint() != r2.props.forceCheckpoint()) {
+			return r1.props.forceCheckpoint() ? -1 : 1;
+		} else if (r1.isPeriodic != r2.isPeriodic) {
+			return r1.isPeriodic ? 1 : -1;
+		} else {
+			return Long.compare(r1.timestamp, r2.timestamp);
+		}
+	});
+
+	CheckpointRequestDecider(
+			int maxConcurrentCheckpointAttempts,
+			Consumer<Long> rescheduleTrigger,
+			Clock clock,
+			long minPauseBetweenCheckpoints,
+			Supplier<Integer> pendingCheckpointsSizeSupplier,
+			Function<Supplier<Optional<CheckpointTriggerRequest>>,
+			Optional<CheckpointTriggerRequest>> sync) {
+		this.maxConcurrentCheckpointAttempts = maxConcurrentCheckpointAttempts;
+		this.rescheduleTrigger = rescheduleTrigger;
+		this.clock = clock;
+		this.minPauseBetweenCheckpoints = minPauseBetweenCheckpoints;
+		this.pendingCheckpointsSizeSupplier = pendingCheckpointsSizeSupplier;
+		this.sync = sync;
+	}
+
+	Optional<CheckpointTriggerRequest> chooseRequestToExecute(CheckpointTriggerRequest newRequest, boolean isTriggering, long lastCompletionMs) {
+		return chooseRequestToExecute(false, () -> newRequest, isTriggering, lastCompletionMs);
+	}
+
+	Optional<CheckpointTriggerRequest> chooseQueuedRequestToExecute(boolean isTriggering, long lastCompletionMs) {
+		return chooseRequestToExecute(true, triggerRequestQueue::peek, isTriggering, lastCompletionMs);
+	}
+
+	/**
+	 * Choose the next {@link CheckpointTriggerRequest request} to execute based on the provided candidate and the
+	 * current state. Acquires a lock and may update the state.
+	 * @return request to execute, if any.
+	 */
+	private Optional<CheckpointTriggerRequest> chooseRequestToExecute(
+			boolean isEnqueued,
+			Supplier<CheckpointTriggerRequest> candidateSupplier,
+			boolean isTriggering,
+			long lastCompletionMs) {
+		return sync.apply(() -> {
+			final CheckpointTriggerRequest candidate = candidateSupplier.get();
+			if (candidate == null) {
+				return Optional.empty();
```

Comment:
This check has to be done inside the `synchronized` section to avoid race conditions.

(I assume you meant `Optional.empty()`, not `@Nullable`; I don't see any `@Nullable` there).